src/snowflake/connector/auth/keypair.py

```diff
@@ -102,26 +103,31 @@ def prepare(
 
         now = datetime.utcnow()
 
-        try:
-            private_key = load_der_private_key(
-                data=self._private_key,
-                password=None,
-                backend=default_backend(),
-            )
-        except Exception as e:
-            raise ProgrammingError(
-                msg=f"Failed to load private key: {e}\nPlease provide a valid "
-                "unencrypted rsa private key in DER format as bytes object",
-                errno=ER_INVALID_PRIVATE_KEY,
-            )
+        if isinstance(self._private_key, bytes):
+            try:
+                private_key = load_der_private_key(
+                    data=self._private_key,
+                    password=None,
+                    backend=default_backend(),
+                )
+            except Exception as e:
+                raise ProgrammingError(
+                    msg=f"Failed to load private key: {e}\nPlease provide a valid "
+                    "unencrypted rsa private key in DER format as bytes object",
+                    errno=ER_INVALID_PRIVATE_KEY,
+                )
 
-        if not isinstance(private_key, RSAPrivateKey):
-            raise ProgrammingError(
-                msg=f"Private key type ({private_key.__class__.__name__}) not supported."
-                "\nPlease provide a valid rsa private key in DER format as bytes "
-                "object",
-                errno=ER_INVALID_PRIVATE_KEY,
-            )
+            if not isinstance(private_key, RSAPrivateKey):
+                raise ProgrammingError(
+                    msg=f"Private key type ({private_key.__class__.__name__}) not supported."
+                    "\nPlease provide a valid rsa private key in DER format as bytes "
+                    "object",
+                    errno=ER_INVALID_PRIVATE_KEY,
+                )
+        elif isinstance(self._private_key, RSAPrivateKey):
+            private_key = self._private_key
+        else:
+            raise TypeError(self._private_key)
```

Comment:
nit: let's be more explicit on the error message: `f"Expected bytes or RSAPrivateKey, got {type(self._private_key)}"`

Reply:
Fixed in f779da2a027a5e9892a87597dc64bfce89fc6401.

Reply:
Let's add an unhappy path test to cover this :) Otherwise LGTM!

Reply:
Fixed in 48e5aca5070353b3594513081e4c6656fe953bd4.